src/common/datomish/transact.cljc

```diff
@@ -552,12 +580,99 @@
     (:db-after (<? (<with db tx-data)))))
 
 (defn <transact!
-  [conn tx-data]
+  "Submits a transaction to the database for writing.
+
+  Returns a pair-chan resolving to `[result error]`."
+  ([conn tx-data]
+   (<transact! conn tx-data (a/chan 1) true))
+  ([conn tx-data result close?]
+   {:pre [(conn? conn)]}
+   ;; Any race to put! is a real race between callers of <transact!.  We can't just park on put!,
+   ;; because the parked putter that is woken is non-deterministic.
+   (let [closed? (not (a/put! (:transact-chan conn) [:sentinel-transact tx-data result close?]))]
+     (go-pair
+       ;; We want to return a pair-chan, no matter what kind of channel result is.
+       (if closed?
+         (raise "Connection is closed" {:error :transact/connection-closed})
+         (<? result))))))
+
+(defn- start-transactor [conn]
+  (let [token-chan (a/chan 1)]
+    (go
+      (>! token-chan (gensym "transactor-token"))
+      (loop []
+        (when-let [token (<! token-chan)]
+          (when-let [[sentinel tx-data result close?] (<! (:transact-chan conn))]
+            (let [pair
+                  (<! (go-pair ;; Catch exceptions, return the pair.
+                        (case sentinel
+                          :sentinel-close
+                          ;; Time to close the underlying DB.
+                          (<? (db/close-db @(:current-db conn)))
+
+                          ;; Default: process the transaction.
+                          (do
+                            (when @(:closed? conn)
+                              ;; Drain enqueued transactions.
+                              (raise "Connection is closed" {:error :transact/connection-closed}))
+                            (let [db (db conn)
+                                  report (<? (db/in-transaction!
+                                               db
+                                               #(-> (<with db tx-data))))]
+                              (when report
+                                ;; <with returns non-nil or throws, but we still check report just in
+                                ;; case.  Here, in-transaction! function completed and returned non-nil,
+                                ;; so the transaction has committed.
+                                (reset! (:current-db conn) (:db-after report))
+                                (>! (:listener-source conn) report))
+                              report)))))]
+              ;; Even when report is nil (transaction not committed), pair is non-nil.
+              (>! result pair))
+            (>! token-chan token)
+            (when close?
+              (a/close! result))
+            (recur)))))))
```

Comment:
You're right, but all that means is that when we park on the channel, we stay there forever, because the channel's closed!

We'll see the last item, process it, recur, get back to line 605, and park forever. We need a check somewhere to decide to terminate, right?